{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example: Introduction to topsbm\n",
    "\n",
    "Topic modelling with hierarchical stochastic block models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2019-01-16T04:07:14.579Z"
    }
   },
   "outputs": [],
   "source": [
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "import pandas as pd\n",
    "from topsbm import TopSBM"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Setup: Load a corpus\n",
    "\n",
    "1. We have a list of documents, each document contains a list of words.\n",
    "1. We have a list of document titles (optional)\n",
    "\n",
    "The example corpus consists of 63 articles from Wikipedia taken from 3 different categories (Experimental Physics, Chemical Physics, and Computational Biology).\n",
    "\n",
    "We use scikit-learn's [`CountVectorizer`](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html) to turn this text into a feature matrix."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2019-01-16T04:06:47.262Z"
    }
   },
   "outputs": [],
   "source": [
    "# Load texts and vectorize\n",
    "with open('corpus.txt', 'r') as f:\n",
    "    docs = f.readlines()\n",
    "\n",
    "vec = CountVectorizer(token_pattern=r'\\S+')\n",
    "X = vec.fit_transform(docs)\n",
    "\n",
    "# X is now a sparse matrix of (docs, words)\n",
    "\n",
    "# titles corresponding to docs\n",
    "with open('titles.txt', 'r') as f:\n",
    "    x = f.readlines()\n",
    "titles = [h.split()[0] for h in x]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-20T11:55:45.791938Z",
     "start_time": "2018-12-20T11:55:45.769609Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nuclear_Overhauser_effect\n",
      " the nuclear overhauser effect noe is the transfer of nuclear spin polarization from one nuclear spi\n"
     ]
    }
   ],
   "source": [
    "# view the data for document 0\n",
    "print(titles[0])\n",
    "print(docs[0][:100])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fit the model\n",
    "\n",
    "Calling `TopSBM.fit_transform` will:\n",
    "* construct the bipartite graph between documents and words (samples and features)\n",
    "* perform Hierarchical Stochastic Block Model inference over the graph\n",
    "* return an embedding of the samples in the block level with finest granularity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-23T00:35:43.395785Z",
     "start_time": "2018-12-22T23:50:22.496049Z"
    }
   },
   "outputs": [],
   "source": [
    "model = TopSBM(weighted_edges=False, random_state=8)\n",
    "Xt = model.fit_transform(X)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plotting the graph and block structure\n",
    "\n",
    "The following plot shows the (hierarchical) community structure in the word-document network as inferred by the stochastic block model:\n",
    "\n",
    "* document-nodes are on the left\n",
    "* word-nodes are on the right\n",
    "* different colors correspond to the different groups\n",
    "\n",
    "The result is a grouping of nodes into groups on multiple levels in the hierarchy:\n",
    "\n",
    "* on the uppermost level, each node belongs to the same group (square in the middle)\n",
    "* on the next-lower level, we split the network into two groups: the word-nodes and the document-nodes (blue sqaures to the left and right, respectively). This is a trivial structure due to the bipartite character of the network.\n",
    "* only next lower levels constitute a non-trivial structure: We now further divide nodes into smaller groups (document-nodes into document-groups on the left and word-nodes into word-groups on the right)\n",
    "\n",
    "In the code, the lowest level is known as level 0, with coarser levels 1, 2, ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-12-20T11:56:56.779616Z",
     "start_time": "2018-12-20T11:56:54.468935Z"
    }
   },
   "outputs": [],
   "source": [
    "model.plot_graph(n_edges=1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Topics\n",
    "\n",
    "For each word-group on a given level in the hierarchy, we retrieve the $n$ most common words in each group -- these are the topics!\n",
    "\n",
    "<span style=\"font-size: large; color: red; font-weight: bold\">Something looks broken here! Are we indexing correctly? Is the graph constructed correctly? Is the inference broken? There is an alarmingly high alphabetic/order correlation in the topic assignment for some topics.</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "topics = pd.DataFrame(model.groups_[1]['p_w_tw'],\n",
    "                      index=vec.get_feature_names())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "anisotropic    0.180791\n",
      "analyzing      0.102034\n",
      "an             0.057401\n",
      "along          0.055311\n",
      "anderson       0.045819\n",
      "attained       0.022542\n",
      "aligned        0.018249\n",
      "acoustic       0.017458\n",
      "american       0.016158\n",
      "algorithms     0.015650\n",
      "Name: 0, dtype: float64\n",
      "\n",
      "bioinformaticians    0.070905\n",
      "assignment           0.033289\n",
      "before               0.020306\n",
      "analyzed             0.015646\n",
      "exclusion            0.015313\n",
      "america              0.013648\n",
      "biodatomics          0.012650\n",
      "air                  0.011651\n",
      "encode               0.011651\n",
      "angles               0.011318\n",
      "Name: 1, dtype: float64\n",
      "\n",
      "transcriptomes    0.090444\n",
      "packages          0.079067\n",
      "linpack           0.072241\n",
      "oxidation         0.026735\n",
      "ozone             0.026166\n",
      "simultaneously    0.023891\n",
      "image             0.023322\n",
      "giving            0.022184\n",
      "showing           0.022184\n",
      "abnormal          0.019340\n",
      "Name: 2, dtype: float64\n",
      "\n",
      "assembled       0.051798\n",
      "alternative     0.041600\n",
      "detail          0.023349\n",
      "decreases       0.022276\n",
      "addressed       0.021739\n",
      "agreement       0.021471\n",
      "acceleration    0.020666\n",
      "characterize    0.017445\n",
      "correct         0.017177\n",
      "acids           0.015835\n",
      "Name: 3, dtype: float64\n",
      "\n",
      "configuration    0.026034\n",
      "extent           0.022315\n",
      "offers           0.021850\n",
      "framed           0.015807\n",
      "dynamics         0.015342\n",
      "continued        0.014412\n",
      "create           0.013947\n",
      "even             0.013482\n",
      "formulated       0.012552\n",
      "hit              0.011158\n",
      "Name: 4, dtype: float64\n",
      "\n",
      "alterations    0.260264\n",
      "although       0.099281\n",
      "assisting      0.011134\n",
      "allowed        0.010670\n",
      "group          0.008583\n",
      "circular       0.007887\n",
      "distributed    0.006727\n",
      "emission       0.006263\n",
      "non            0.006263\n",
      "breaking       0.005567\n",
      "Name: 5, dtype: float64\n",
      "\n",
      "accurately      0.029478\n",
      "academic        0.029104\n",
      "basic           0.025373\n",
      "chaperones      0.024627\n",
      "affiliated      0.022015\n",
      "austria         0.021269\n",
      "entered         0.020149\n",
      "abyss           0.018657\n",
      "assays          0.017537\n",
      "coefficients    0.017537\n",
      "Name: 6, dtype: float64\n",
      "\n",
      "coherent     0.036683\n",
      "apl          0.031898\n",
      "exterior     0.030303\n",
      "chair        0.029771\n",
      "brains       0.028177\n",
      "being        0.023392\n",
      "model        0.022860\n",
      "formation    0.022329\n",
      "mid          0.021797\n",
      "binding      0.020734\n",
      "Name: 7, dtype: float64\n",
      "\n",
      "glasgow       0.082265\n",
      "sites         0.077991\n",
      "coil          0.070513\n",
      "longer        0.054487\n",
      "kb            0.053419\n",
      "huntingtin    0.048077\n",
      "chang         0.035256\n",
      "effect        0.033120\n",
      "medium        0.028846\n",
      "momentum      0.027778\n",
      "Name: 8, dtype: float64\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for topic in topics.columns:\n",
    "    print(topics[topic].nlargest(10))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Topic-distribution in each document\n",
    "\n",
    "Which level-1 topics contribute to each document?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Nuclear_Overhauser_effect</th>\n",
       "      <th>Quantum_solvent</th>\n",
       "      <th>Rovibrational_coupling</th>\n",
       "      <th>Effective_field_theory</th>\n",
       "      <th>Chemical_physics</th>\n",
       "      <th>Rotational_transition</th>\n",
       "      <th>Dynamic_nuclear_polarisation</th>\n",
       "      <th>Knight_shift</th>\n",
       "      <th>Polarizability</th>\n",
       "      <th>Anisotropic_liquid</th>\n",
       "      <th>...</th>\n",
       "      <th>Louis_and_Beatrice_Laufer_Center_for_Physical_and_Quantitative_Biology</th>\n",
       "      <th>Law_of_Maximum</th>\n",
       "      <th>Enzyme_Function_Initiative</th>\n",
       "      <th>SnoRNA_prediction_software</th>\n",
       "      <th>Sepp_Hochreiter</th>\n",
       "      <th>Aureus_Sciences</th>\n",
       "      <th>IEEE/ACM_Transactions_on_Computational_Biology_and_Bioinformatics</th>\n",
       "      <th>Knotted_protein</th>\n",
       "      <th>BioUML</th>\n",
       "      <th>De_novo_transcriptome_assembly</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.460145</td>\n",
       "      <td>0.538136</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.530474</td>\n",
       "      <td>0.478723</td>\n",
       "      <td>0.478992</td>\n",
       "      <td>0.434272</td>\n",
       "      <td>0.481481</td>\n",
       "      <td>0.478170</td>\n",
       "      <td>0.514286</td>\n",
       "      <td>...</td>\n",
       "      <td>0.357143</td>\n",
       "      <td>0.531469</td>\n",
       "      <td>0.362044</td>\n",
       "      <td>0.517241</td>\n",
       "      <td>0.391371</td>\n",
       "      <td>0.348485</td>\n",
       "      <td>0.340909</td>\n",
       "      <td>0.451087</td>\n",
       "      <td>0.409978</td>\n",
       "      <td>0.432133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.268116</td>\n",
       "      <td>0.161017</td>\n",
       "      <td>0.268018</td>\n",
       "      <td>0.076749</td>\n",
       "      <td>0.138298</td>\n",
       "      <td>0.168067</td>\n",
       "      <td>0.330986</td>\n",
       "      <td>0.343915</td>\n",
       "      <td>0.262994</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>...</td>\n",
       "      <td>0.010204</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.008759</td>\n",
       "      <td>0.017241</td>\n",
       "      <td>0.004622</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016304</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.005540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.003623</td>\n",
       "      <td>0.004237</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000520</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.040816</td>\n",
       "      <td>0.006993</td>\n",
       "      <td>0.128467</td>\n",
       "      <td>0.086207</td>\n",
       "      <td>0.043143</td>\n",
       "      <td>0.121212</td>\n",
       "      <td>0.022727</td>\n",
       "      <td>0.070652</td>\n",
       "      <td>0.078091</td>\n",
       "      <td>0.114497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.079710</td>\n",
       "      <td>0.067797</td>\n",
       "      <td>0.020270</td>\n",
       "      <td>0.038375</td>\n",
       "      <td>0.053191</td>\n",
       "      <td>0.033613</td>\n",
       "      <td>0.045775</td>\n",
       "      <td>0.031746</td>\n",
       "      <td>0.027547</td>\n",
       "      <td>0.085714</td>\n",
       "      <td>...</td>\n",
       "      <td>0.224490</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.268613</td>\n",
       "      <td>0.258621</td>\n",
       "      <td>0.331279</td>\n",
       "      <td>0.227273</td>\n",
       "      <td>0.318182</td>\n",
       "      <td>0.255435</td>\n",
       "      <td>0.253796</td>\n",
       "      <td>0.278855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.032609</td>\n",
       "      <td>0.033898</td>\n",
       "      <td>0.011261</td>\n",
       "      <td>0.027088</td>\n",
       "      <td>0.021277</td>\n",
       "      <td>0.016807</td>\n",
       "      <td>0.018779</td>\n",
       "      <td>0.015873</td>\n",
       "      <td>0.012994</td>\n",
       "      <td>0.114286</td>\n",
       "      <td>...</td>\n",
       "      <td>0.030612</td>\n",
       "      <td>0.027972</td>\n",
       "      <td>0.011679</td>\n",
       "      <td>0.017241</td>\n",
       "      <td>0.016949</td>\n",
       "      <td>0.015152</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016304</td>\n",
       "      <td>0.017354</td>\n",
       "      <td>0.014774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.061594</td>\n",
       "      <td>0.076271</td>\n",
       "      <td>0.051802</td>\n",
       "      <td>0.040632</td>\n",
       "      <td>0.074468</td>\n",
       "      <td>0.025210</td>\n",
       "      <td>0.056338</td>\n",
       "      <td>0.026455</td>\n",
       "      <td>0.054574</td>\n",
       "      <td>0.028571</td>\n",
       "      <td>...</td>\n",
       "      <td>0.142857</td>\n",
       "      <td>0.062937</td>\n",
       "      <td>0.134307</td>\n",
       "      <td>0.086207</td>\n",
       "      <td>0.098613</td>\n",
       "      <td>0.121212</td>\n",
       "      <td>0.113636</td>\n",
       "      <td>0.119565</td>\n",
       "      <td>0.151844</td>\n",
       "      <td>0.105263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.083333</td>\n",
       "      <td>0.101695</td>\n",
       "      <td>0.119369</td>\n",
       "      <td>0.232506</td>\n",
       "      <td>0.085106</td>\n",
       "      <td>0.151261</td>\n",
       "      <td>0.099765</td>\n",
       "      <td>0.095238</td>\n",
       "      <td>0.138773</td>\n",
       "      <td>0.171429</td>\n",
       "      <td>...</td>\n",
       "      <td>0.030612</td>\n",
       "      <td>0.048951</td>\n",
       "      <td>0.024818</td>\n",
       "      <td>0.017241</td>\n",
       "      <td>0.053929</td>\n",
       "      <td>0.030303</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027174</td>\n",
       "      <td>0.010846</td>\n",
       "      <td>0.014774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.010870</td>\n",
       "      <td>0.016949</td>\n",
       "      <td>0.029279</td>\n",
       "      <td>0.051919</td>\n",
       "      <td>0.138298</td>\n",
       "      <td>0.126050</td>\n",
       "      <td>0.010563</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.023389</td>\n",
       "      <td>0.057143</td>\n",
       "      <td>...</td>\n",
       "      <td>0.081633</td>\n",
       "      <td>0.244755</td>\n",
       "      <td>0.007299</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.009245</td>\n",
       "      <td>0.030303</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.016304</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002257</td>\n",
       "      <td>0.010638</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.003521</td>\n",
       "      <td>0.005291</td>\n",
       "      <td>0.001040</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.081633</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.054015</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.050847</td>\n",
       "      <td>0.090909</td>\n",
       "      <td>0.204545</td>\n",
       "      <td>0.027174</td>\n",
       "      <td>0.078091</td>\n",
       "      <td>0.034164</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows × 63 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Nuclear_Overhauser_effect  Quantum_solvent  Rovibrational_coupling  \\\n",
       "0                   0.460145         0.538136                0.500000   \n",
       "1                   0.268116         0.161017                0.268018   \n",
       "2                   0.003623         0.004237                0.000000   \n",
       "3                   0.079710         0.067797                0.020270   \n",
       "4                   0.032609         0.033898                0.011261   \n",
       "5                   0.061594         0.076271                0.051802   \n",
       "6                   0.083333         0.101695                0.119369   \n",
       "7                   0.010870         0.016949                0.029279   \n",
       "8                   0.000000         0.000000                0.000000   \n",
       "\n",
       "   Effective_field_theory  Chemical_physics  Rotational_transition  \\\n",
       "0                0.530474          0.478723               0.478992   \n",
       "1                0.076749          0.138298               0.168067   \n",
       "2                0.000000          0.000000               0.000000   \n",
       "3                0.038375          0.053191               0.033613   \n",
       "4                0.027088          0.021277               0.016807   \n",
       "5                0.040632          0.074468               0.025210   \n",
       "6                0.232506          0.085106               0.151261   \n",
       "7                0.051919          0.138298               0.126050   \n",
       "8                0.002257          0.010638               0.000000   \n",
       "\n",
       "   Dynamic_nuclear_polarisation  Knight_shift  Polarizability  \\\n",
       "0                      0.434272      0.481481        0.478170   \n",
       "1                      0.330986      0.343915        0.262994   \n",
       "2                      0.000000      0.000000        0.000520   \n",
       "3                      0.045775      0.031746        0.027547   \n",
       "4                      0.018779      0.015873        0.012994   \n",
       "5                      0.056338      0.026455        0.054574   \n",
       "6                      0.099765      0.095238        0.138773   \n",
       "7                      0.010563      0.000000        0.023389   \n",
       "8                      0.003521      0.005291        0.001040   \n",
       "\n",
       "   Anisotropic_liquid               ...                \\\n",
       "0            0.514286               ...                 \n",
       "1            0.028571               ...                 \n",
       "2            0.000000               ...                 \n",
       "3            0.085714               ...                 \n",
       "4            0.114286               ...                 \n",
       "5            0.028571               ...                 \n",
       "6            0.171429               ...                 \n",
       "7            0.057143               ...                 \n",
       "8            0.000000               ...                 \n",
       "\n",
       "   Louis_and_Beatrice_Laufer_Center_for_Physical_and_Quantitative_Biology  \\\n",
       "0                                           0.357143                        \n",
       "1                                           0.010204                        \n",
       "2                                           0.040816                        \n",
       "3                                           0.224490                        \n",
       "4                                           0.030612                        \n",
       "5                                           0.142857                        \n",
       "6                                           0.030612                        \n",
       "7                                           0.081633                        \n",
       "8                                           0.081633                        \n",
       "\n",
       "   Law_of_Maximum  Enzyme_Function_Initiative  SnoRNA_prediction_software  \\\n",
       "0        0.531469                    0.362044                    0.517241   \n",
       "1        0.000000                    0.008759                    0.017241   \n",
       "2        0.006993                    0.128467                    0.086207   \n",
       "3        0.076923                    0.268613                    0.258621   \n",
       "4        0.027972                    0.011679                    0.017241   \n",
       "5        0.062937                    0.134307                    0.086207   \n",
       "6        0.048951                    0.024818                    0.017241   \n",
       "7        0.244755                    0.007299                    0.000000   \n",
       "8        0.000000                    0.054015                    0.000000   \n",
       "\n",
       "   Sepp_Hochreiter  Aureus_Sciences  \\\n",
       "0         0.391371         0.348485   \n",
       "1         0.004622         0.015152   \n",
       "2         0.043143         0.121212   \n",
       "3         0.331279         0.227273   \n",
       "4         0.016949         0.015152   \n",
       "5         0.098613         0.121212   \n",
       "6         0.053929         0.030303   \n",
       "7         0.009245         0.030303   \n",
       "8         0.050847         0.090909   \n",
       "\n",
       "   IEEE/ACM_Transactions_on_Computational_Biology_and_Bioinformatics  \\\n",
       "0                                           0.340909                   \n",
       "1                                           0.000000                   \n",
       "2                                           0.022727                   \n",
       "3                                           0.318182                   \n",
       "4                                           0.000000                   \n",
       "5                                           0.113636                   \n",
       "6                                           0.000000                   \n",
       "7                                           0.000000                   \n",
       "8                                           0.204545                   \n",
       "\n",
       "   Knotted_protein    BioUML  De_novo_transcriptome_assembly  \n",
       "0         0.451087  0.409978                        0.432133  \n",
       "1         0.016304  0.000000                        0.005540  \n",
       "2         0.070652  0.078091                        0.114497  \n",
       "3         0.255435  0.253796                        0.278855  \n",
       "4         0.016304  0.017354                        0.014774  \n",
       "5         0.119565  0.151844                        0.105263  \n",
       "6         0.027174  0.010846                        0.014774  \n",
       "7         0.016304  0.000000                        0.000000  \n",
       "8         0.027174  0.078091                        0.034164  \n",
       "\n",
       "[9 rows x 63 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(model.groups_[1]['p_tw_d'],\n",
    "             columns=titles)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extra: Clustering of documents - for free.\n",
    "\n",
    "The stochastic block models clusters the documents into groups. We do not need to run an additional clustering to obtain this grouping.\n",
    "\n",
    "For a query article, we can return all articles from the same group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Nuclear_Overhauser_effect                        0\n",
       "Quantum_solvent                                  0\n",
       "Rovibrational_coupling                           0\n",
       "Effective_field_theory                           0\n",
       "Chemical_physics                                 0\n",
       "Rotational_transition                            0\n",
       "Dynamic_nuclear_polarisation                     0\n",
       "Knight_shift                                     0\n",
       "Polarizability                                   0\n",
       "Anisotropic_liquid                               0\n",
       "Rotating_wave_approximation                      0\n",
       "RRKM_theory                                      0\n",
       "Molecular_vibration                              0\n",
       "Electrostatic_deflection_(structural_element)    0\n",
       "Magic_angle_(EELS)                               0\n",
       "Reactive_empirical_bond_order                    0\n",
       "McConnell_equation                               0\n",
       "Ziff-Gulari-Barshad_model                        0\n",
       "Empirical_formula                                0\n",
       "Pauli_effect                                     0\n",
       "Ripple_tank                                      0\n",
       "Complementary_experiments                        0\n",
       "Wave_tank                                        0\n",
       "X-ray_crystal_truncation_rod                     0\n",
       "X-ray_standing_waves                             0\n",
       "Fragment_separator                               0\n",
       "Dynamic_mode_decomposition                       0\n",
       "Quantum_oscillations_(experimental_technique)    0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster_labels = pd.DataFrame(model.groups_[1]['p_td_d'],\n",
    "                              columns=titles).idxmax(axis=0)\n",
    "cluster_idx = cluster_labels['Rovibrational_coupling']\n",
    "cluster_labels[cluster_labels == cluster_idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
